ampl-converter/src/main/java/com/powsybl/ampl/converter/AmplNetworkReader.java

```diff
@@ -78,60 +81,69 @@ public AmplNetworkReader readGenerators() throws IOException {
             g.setTargetV(g.getTerminal().getVoltageLevel().getNominalV());
         }
 
-        read("_generators", 4, this::readGenerator);
+        read("_generators", 8, this::readGenerator);
 
         return this;
     }
 
     private Void readGenerator(String[] tokens) {
         int num = Integer.parseInt(tokens[0]);
-        float v = Float.parseFloat(tokens[1]);
-        float p = Float.parseFloat(tokens[2]);
-        float q = Float.parseFloat(tokens[3]);
+        int busNum = tokens[1] != null ? Integer.parseInt(tokens[1]) : -1;
+        boolean vregul = Boolean.parseBoolean(tokens[2]);
+        float targetV = Float.parseFloat(tokens[3]);
+        float targetP = Float.parseFloat(tokens[4]);
+        float targetQ = Float.parseFloat(tokens[5]);
+        float p = Float.parseFloat(tokens[6]);
+        float q = Float.parseFloat(tokens[7]);
         String id = mapper.getId(AmplSubset.GENERATOR, num);
         Generator g = network.getGenerator(id);
         if (g == null) {
             throw new AmplException("Invalid generator id '" + id + "'");
         }
+
+        g.setVoltageRegulatorOn(vregul);
+        g.setTargetV(targetV);
+        g.setTargetP(targetP);
+        g.setTargetQ(targetQ);
+
         Terminal t = g.getTerminal();
-        float vb = t.getVoltageLevel().getNominalV();
-        g.setTargetV(v * vb);
-        g.setTargetP(p);
-        g.setTargetQ(q);
-        g.getTerminal().setP(-p).setQ(-q);
-        if (t.isConnected()) {
-            if (p == 0 && q == 0) {
+        t.setP(p).setQ(q);
+
+        if (busNum == -1) {
+            if (t.isConnected()) {
                 t.disconnect();
             }
         } else {
-            if (p != 0 || q != 0) {
+            String busId = mapper.getId(AmplSubset.BUS, busNum);
+            if (t.getConnectable().getId().equals(busId)) {
```

Comment:
I have a doubt about this line (and the other duplicated from this) because here you compare t.getConnectable() which is an equipment and a bus.